plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java

```diff
@@ -156,33 +162,39 @@ static String dateToStringConverter(Object value)
 
     static String datetimeToStringConverter(Object value)
     {
-        return formatTimestamp(divide(((long) value), MICROSECONDS_PER_MILLISECOND, RoundingMode.UNNECESSARY), UTC);
+        long epochMicros = (long) value;
+        long epochSeconds = floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
+        int nanoAdjustment = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+        return formatTimestamp(epochSeconds, nanoAdjustment, UTC);
     }
 
     static String timeToStringConverter(Object value)
     {
-        long longValue = ((Long) value).longValue();
-        long nanosUtc = DateTimeEncoding.unpackTimeNanos(longValue);
-        ZoneId zoneId = ZoneId.of(DateTimeEncoding.unpackZoneKey(longValue).getId());
-        return TIME_FORMATTER.format(toZonedDateTime(nanosUtc / NANOSECONDS_PER_MILLISECOND, zoneId));
+        long time = (long) value;
+        long epochSeconds = floorDiv(time, PICOSECONDS_PER_SECOND);
+        long nanoAdjustment = floorDiv(floorMod(time, PICOSECONDS_PER_SECOND), PICOSECONDS_PER_NANOSECOND);
```

Comment:
- no need to use `floorDiv` and `floorMod`, since `value` should be non-negative. (You can always `verify(0 <= time, "Invalid time value: %s", time);`)
- the additional `floorDiv` (in `floorDiv(floorMod`) is redundant anyway, since floorMod returns a positive value.  Use ordinary `/` for this one too